sb3_contrib/ars/ars.py

```diff
@@ -0,0 +1,276 @@
+
+            callback.on_rollout_start()
+            while theta_idx < candidate_returns.shape[0]:
+                policy_list = []
+
+                # We are using a vecenv with n_envs==n_workers. We batch our candidate theta evaluations into vectors
+                # of length n_workers
+                for _ in range(self.n_workers):
+                    theta_tensor = th.tensor(candidate_thetas[theta_idx], dtype=self.dtype)
+                    th.nn.utils.vector_to_parameters(theta_tensor, self.policy.parameters())
+                    policy_list.append(copy.deepcopy(self.policy))
+                    theta_idx += 1
+
+                # We only want one rollout per policy, but vec envs will automatically reset when a done is sent.
+                # To get around this we use vec_done_once to keep track of which environments have already finished,
+                # and we ignore rewards once we see the first done.
+                vec_obs = self.env.reset()
+                vec_acts = np.zeros((self.n_workers, self.action_space.shape[0]))
+                vec_returns = np.zeros(self.n_workers)
+                vec_done_once = np.zeros(self.n_workers, dtype=bool)  # Tracks if each env has returned done at least once
+                vec_steps_taken = np.zeros(self.n_workers)  # used to update logs
+
+                while not np.all(vec_done_once):
+                    for i in range(self.n_workers):
+                        vec_acts[i, :] = policy_list[i].predict(vec_obs[i, :])[0]
+                    vec_obs, vec_rews, vec_done, vec_info = self.env.step(vec_acts)
+                    callback.on_step()
+
+                    vec_rews += self.alive_bonus_offset  # Alive offset from the original paper, defaults to zero
+
+                    # The order of episodes in the ep_info_buffer won't match up with the order of theta, which feels gross,
+                    # But we are only using ep_info for logging so I don't think that should natter
+                    for i, (done_this_time, done_before) in enumerate(zip(vec_done, vec_done_once)):
+                        if done_this_time and not done_before:
+                            maybe_ep_info = vec_info[i].get("episode")
+                            if maybe_ep_info is not None:
+                                self.ep_info_buffer.append(maybe_ep_info)
+
+                    vec_steps_taken += np.invert(vec_done_once)
+                    vec_returns += vec_rews * np.invert(vec_done_once)
+                    vec_done_once = np.bitwise_or(vec_done, vec_done_once)
+
+                candidate_returns[theta_idx - self.n_workers : theta_idx] = vec_returns
+                batch_steps += sum(vec_steps_taken)  # only count steps used in the update step towards our total_steps
+
+            callback.on_rollout_end()
+
+        return candidate_returns, batch_steps
+
+    # Make sure our hyper parameters are valid and auto correct them if they are not
+    def _validate_hypers(self):
```

Comment:
Avoid unnecessary shorthands (change to `_validate_hyperparameters` or `_validate_hyperparams`).